[inductor][choices] lookup table choices 1/3

coconutruben opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #164980
* #164979
* __->__ #164978

\# why

- enable users to control which choices get used on which inputs
- reduce lowering time, and pin kernel selection, by selecting
  them for the inputs

\# what

- a new InductorChoices subclass that implements a lookup table
- a README explaining the usage
- corresponding testing

- currently only supports templates that go through
  `V.choices.get_template_configs`

\# testing

```
python3 -bb -m pytest test/inductor/test_lookup_table.py -v
```

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov @mlazos

Differential Revision: [D85685743](https://our.internmc.facebook.com/intern/diff/D85685743)

coconutruben committed:
[inductor][choices] lookup table choices 1/3

\# why

- enable users to control which choices get used on which inputs
- reduce lowering time, and pin kernel selection, by selecting
  them for the inputs

\# what

- a new InductorChoices subclass that implements a lookup table
- a README explaining the usage
- corresponding testing

- currently only supports templates that go through
  `V.choices.get_template_configs`

\# testing

```
python3 -bb -m pytest test/inductor/test_lookup_table.py -v
```

[ghstack-poisoned]

PaulZhang12 reviewed (approved):
LGTM! cc @jansel @eellison

eellison commented on torch/_inductor/lookup_table/README.md:
````diff
@@ -0,0 +1,153 @@
+```
+
+**Lookup Behavior**: During lookup, the system automatically tries both key formats:
+1. **Device-specific key** (e.g., `"NVIDIA H100+input_data+mm"`) - tried first
````
nit: is there any specification of the key format?

coconutruben replied:
I added a key schema section and a comment explaining the device specific key and encouraging the user to file a request if they need a CUDA specific key. device key is really only there to help deployments, rather than needing BUCK rules or other fancy things to swap out the table from the codebase depending on the target architecture - there are probably multitude of those use cases, I think I'd be more helpful for users to ask for them rather than building them proactively (e.g. can you have a cuda version without a specific device, etc)

eellison commented on torch/_inductor/lookup_table/README.md:
```diff
@@ -0,0 +1,153 @@
+## Implementation Details
+
+### Key Generation
+- Device key: Uses `torch.cuda.get_device_properties().gcnArchName` (e.g., "NVIDIA H100")
```
I guess this is the key specification part. I think this is fine if it's a bit underspecified, and subject to change.

eellison commented on torch/_inductor/lookup_table/choices.py:
```diff
@@ -0,0 +1,391 @@
+        Override this method to use custom lookup table sources (database, API, etc.).
+        """
+        if not torch.cuda.is_available() or config.lookup_table.table is None:
+            return {}
```
nit: in the future, no specific need for `torch.cuda.is_available`

PaulZhang12 commented on torch/_inductor/lookup_table/choices.py:
```diff
@@ -0,0 +1,391 @@
+        # Get node information using existing methods
+        dtypes = kernel_inputs.dtypes()
+        shapes = kernel_inputs.shapes_hinted()
+        strides = kernel_inputs.strides_hinted()
```
cc @aorenste these are cases that we probably want to test for in distributed autotuning as well? Maybe this can be a handoff

coconutruben committed:
Update on "[inductor][choices] lookup table choices 1/3"


\# why

- enable users to control which choices get used on which inputs
- reduce lowering time, and pin kernel selection, by selecting
  them for the inputs

\# what

- a new InductorChoices subclass that implements a lookup table
- a README explaining the usage
- corresponding testing

- currently only supports templates that go through
  `V.choices.get_template_configs`

\# testing

```
python3 -bb -m pytest test/inductor/test_lookup_table.py -v
```

[ghstack-poisoned]

malfet commented:
@pytorchbot revert -m "Looks like it broke slow tests, see https://hud.pytorch.org/hud/pytorch/pytorch/cbcb4f776835dfd8350b70a6ffb52503679241b2/1?per_page=50&name_filter=slow&mergeEphemeralLF=true" -c nosignal

coconutruben committed:
Update on "[inductor][choices] lookup table choices 1/3"


\# why

- enable users to control which choices get used on which inputs
- reduce lowering time, and pin kernel selection, by selecting
  them for the inputs

\# what

- a new InductorChoices subclass that implements a lookup table
- a README explaining the usage
- corresponding testing

- currently only supports templates that go through
  `V.choices.get_template_configs`

\# testing

```
python3 -bb -m pytest test/inductor/test_lookup_table.py -v
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov mlazos

[ghstack-poisoned]

coconutruben committed:
Update on "[inductor][choices] lookup table choices 1/3"


\# why

- enable users to control which choices get used on which inputs
- reduce lowering time, and pin kernel selection, by selecting
  them for the inputs

\# what

- a new InductorChoices subclass that implements a lookup table
- a README explaining the usage
- corresponding testing

- currently only supports templates that go through
  `V.choices.get_template_configs`

\# testing

```
python3 -bb -m pytest test/inductor/test_lookup_table.py -v
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov mlazos

[ghstack-poisoned]

coconutruben commented:
@pytorchbot merge

atalman commented:
@pytorchmergebot revert -c nosignal -m "failing internal test on newly added tests: Test when there's no lookup table entry with different autotune modes"

coconutruben commented:
D85812420 should have mitigated these issues internally

@pytorchbot merge
